cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -426,9 +438,52 @@ public Set<ProgramId> filterProgramsExistence(Collection<ProgramId> programIds)
   }
 
 
-  public void writeApplication(String namespaceId, String appId, String versionId, ApplicationSpecification spec)
-    throws IOException {
-    writeApplicationSerialized(namespaceId, appId, versionId, GSON.toJson(new ApplicationMeta(appId, spec)));
+  public void writeApplication(String namespaceId, String appId, String versionId, ApplicationSpecification spec,
+                               ChangeDetail change) throws IOException {
+    writeApplicationSerialized(namespaceId, appId, versionId, GSON.toJson(new ApplicationMeta(appId, spec, change)),
+                               change.getCreationTimeMillis(), change.getAuthor(), change.getDescription());
+  }
+
+  public void createApplicationVersion(ApplicationId id, ApplicationMeta meta)
+    throws IOException, ConflictException {
+    // Fetch the latest version
+    ApplicationMeta latest = getLatest(id.getNamespaceId(), id.getApplication());
+
+    if (!deployAppAllowed(meta.getChange().getParentVersion(), latest)) {
+      throw new ConflictException("Can't deploy the application because the parent version is not the latest.");
```

Comment:
nit: we try to avoid contractions like "can't" in user visible text. It also would be good to include the versions in the error message. Something like:

Cannot deploy the application because parent version '[parent version]' does not match the latest version '[latest version]'